### PR TITLE
Allow opt-out of PII in CONTRIBUTORS.txt

### PR DESCRIPTION
**Pull request by therealbnut**
Update generate_contributors_list.sh to allow removed emails (#2452).

### Motivation:

`CONTRIBUTORS.txt` is leaking PII of contributors and making them easily searchable on sites like Google. The PII remains in the commit history, but that does not seem to be indexed in a similar way. This change adds an opt-out way to have these emails removed.

### Modifications:

This adds a new file `.mailfilter` which has a similar purpose to `.mailmap`, but is used to filter instead of map. The contents of this file are lines with a sha1 hash of each opted-out contributor's name email pair. The hash is intended to obfuscate its entries from search indexing, but does not need to be secure otherwise as the information is still in the git log.

If the git log format ever changes then this would need to be regenerated. Regeneration could easily be done by re-running the script from the same commit it was last generated from, and hashing the diff additions in `CONTRIBUTORS.txt`.

The script `generate_contributors_list.sh` has also been updated to use the filter, although `mailmap` doesn't appear to be used there.

### Result:

`CONTRIBUTORS.txt` can now be anonymised.

```diff
-- Someone's name <the-old-email>
+- Someone's name <email-removed>
```

### Performance:

The bash code isn't the most efficient, but I aimed for a small diff, and compatibility with multiple bash versions. It takes 2 seconds wall-time on my 2021 Apple M1 Max.

I believe the main performance bottleneck is the `O(n log(n))` complexity of the hash lookup. It could be made more efficient with more modern bash's `declare -A`, or by using another language with efficient set membership. I don't think this is run frequently enough for it to be an issue. If it was run on CI, or there was a much larger number of contributors it may be worth exploring a different approach.

### CODE REVIEW

**Comment by therealbnut** [comment]
NOTE: `CONTRIBUTORS.txt` may not have `.mailmap` applied, I'm not sure how that's meant to be applied normally.

**Comment by therealbnut** [comment]
NOTE: After removing all of my entries from `.mailmap` there's now 3 entries for me in `CONTRIBUTORS.txt` some with just my username, some with just my first name. I think the most pragramatic solution is to remove the entry rather than show `<email-removed>`.

**Review by Lukasa — APPROVED** [review]
LGTM, thanks!

**Comment by Lukasa** [comment]
@swift-server-bot test this please

**Comment by therealbnut** [comment]
> LGTM, thanks!

Thank you!

File: CONTRIBUTORS.txt [hunk 1]
```diff
@@ -62,8 +66,10 @@ needs to be listed here.
 - GuangGuang <lihansey163@163.com>
 - Guillaume Lessard <guillaume.lessard@apple.com>
 - Gwynne Raskind <gwynne@darkrainfall.org>
+- Gwynne Raskind <gwynne@vapor.codes>
```
Comment (weissi) [inline CONTRIBUTORS.txt:69]: CC @gwynne you might want to unify your entry using `.mailmap` too?

File: CONTRIBUTORS.txt [hunk 2]
```diff
@@ -97,13 +103,15 @@ needs to be listed here.
...
 - Luo Xiu <luoxiustm@gmail.com>
+- Mahdi Bahrami <github@mahdibm.com>
 - Marcus Liotta <marcus@liotta.se>
 - Mario Sangiorgio <msangiorgio@apple.com>
 - Markus Jais <markusjais@googlemail.com>
 - Markus Kieselmann <markus@kieselmann.eu>
 - Marli Oshlack <marli.oshlack@apple.com>
 - Matt Eaton <agnosticdev@gmail.com>
 - Matteo Comisso <matteo.comisso@me.com>
+- Max Desiatov <m_desiatov@apple.com>
```
Comment (weissi) [inline CONTRIBUTORS.txt:114]: CC @MaxDesiatov you might want to unify your entries using `.mailmap`?

File: CONTRIBUTORS.txt [hunk 3]
```diff
@@ -114,14 +122,18 @@ needs to be listed here.
 - Norman Maurer <norman_maurer@apple.com>
 - Ole Begemann <ole@oleb.net>
 - OleG <emptystamp@gmail.com>
+- Oleksandr Zhurba <69315809+mob-connection@users.noreply.github.com>
```
Comment (weissi) [inline CONTRIBUTORS.txt:125]: CC @mob-connection  if you want a real email address here, please submit a pull request with an appropriate entry to the `.mailmap` file. You'll find plenty of examples there, but essentially if you add

```
Your Name <your-real@email.address> <123456+the-weird-github-one@users.noreply.github.com>
```

and then run `scripts/generate_contributors_list.sh` you should appear with proper name & email.

File: CONTRIBUTORS.txt [hunk 4]
```diff
@@ -114,14 +122,18 @@ needs to be listed here.
...
 - Paul Schmiedmayer <paul.schmiedmayer@tum.de>
 - Peter Adams <pp_adams@apple.com>
 - Peter Tolsma <pejato@umich.edu>
 - Petro Rovenskyy <petro.rovenskyy@gmail.com>
+- Philip Brown <122590765+asdf-bro@users.noreply.github.com>
```
Comment (weissi) [inline CONTRIBUTORS.txt:130]: CC @asdf-bro  if you want a real email address here, please submit a pull request with an appropriate entry to the `.mailmap` file. You'll find plenty of examples there, but essentially if you add

```
Your Name <your-real@email.address> <123456+the-weird-github-one@users.noreply.github.com>
```

and then run `scripts/generate_contributors_list.sh` you should appear with proper name & email.

File: CONTRIBUTORS.txt [hunk 5]
```diff
@@ -114,14 +122,18 @@ needs to be listed here.
...
 - Portia <portia-s@users.noreply.github.com>
 - Pushkar N Kulkarni <pushkar.nk@in.ibm.com>
 - Pushparaj Zala <pr92zala@gmail.com>
 - Richard Low <ral@apple.com>
+- Rick Newton-Rogers <104022490+rnro@users.noreply.github.com>
```
Comment (weissi) [inline CONTRIBUTORS.txt:135]: CC @rnro if you want a real email address here, please submit a pull request with an appropriate entry to the `.mailmap` file. You'll find plenty of examples there, but essentially if you add

```
Your Name <your-real@email.address> <123456+the-weird-github-one@users.noreply.github.com>
```

and then run `scripts/generate_contributors_list.sh` you should appear with proper name & email.

File: CONTRIBUTORS.txt [hunk 6]
```diff
@@ -152,11 +164,20 @@ needs to be listed here.
 - Volodymyr Sapsai <vsapsai@users.noreply.github.com>
 - Will Lisac <will@lisac.org>
 - Wilson Ding <hello@wilsonding.com>
+- YR Chen <stevapple@icloud.com>
+- Yim Lee <yim_lee@apple.com>
 - akash-55 <61596874+akash-55@users.noreply.github.com>
+- buttaface <butta@fastem.com>
+- carolinacass <67160898+carolinacass@users.noreply.github.com>
```
Comment (weissi) [inline CONTRIBUTORS.txt:171]: CC @carolinacass  if you want a real email address here, please submit a pull request with an appropriate entry to the `.mailmap` file. You'll find plenty of examples there, but essentially if you add

```
Your Name <your-real@email.address> <123456+the-weird-github-one@users.noreply.github.com>
```

and then run `scripts/generate_contributors_list.sh` you should appear with proper name & email.

File: CONTRIBUTORS.txt [hunk 7]
```diff
@@ -152,11 +164,20 @@ needs to be listed here.
...
+- dkz2 <132388278+dkz2@users.noreply.github.com>
```
Comment (weissi) [inline CONTRIBUTORS.txt:172]: CC @dkz2  if you want a real email address here, please submit a pull request with an appropriate entry to the `.mailmap` file. You'll find plenty of examples there, but essentially if you add

```
Your Name <your-real@email.address> <123456+the-weird-github-one@users.noreply.github.com>
```

and then run `scripts/generate_contributors_list.sh` you should appear with proper name & email.

File: CONTRIBUTORS.txt [hunk 8]
```diff
@@ -152,11 +164,20 @@ needs to be listed here.
...
 - fadi-botros <botros_fadi@yahoo.com>
 - jemmons <jemmons@users.noreply.github.com>
 - pokryfka <pokryfka@gmail.com>
+- ser <122270051+ser-0xff@users.noreply.github.com>
```
Comment (weissi) [inline CONTRIBUTORS.txt:176]: CC @ser-0xff  if you want a real email address here, please submit a pull request with an appropriate entry to the `.mailmap` file. You'll find plenty of examples there, but essentially if you add

```
Your Name <your-real@email.address> <123456+the-weird-github-one@users.noreply.github.com>
```

and then run `scripts/generate_contributors_list.sh` you should appear with proper name & email.

File: CONTRIBUTORS.txt [hunk 9]
```diff
@@ -152,11 +164,20 @@ needs to be listed here.
...
 - sun_ <sunlubo.sun@gmail.com>
+- taylorswift <kelvin13ma@gmail.com>
+- thomas-gill-0xff <115996764+thomas-gill-0xff@users.noreply.github.com>
```
Comment (weissi) [inline CONTRIBUTORS.txt:179]: CC @thomas-gill-0xff  if you want a real email address here, please submit a pull request with an appropriate entry to the `.mailmap` file. You'll find plenty of examples there, but essentially if you add

```
Your Name <your-real@email.address> <123456+the-weird-github-one@users.noreply.github.com>
```

and then run `scripts/generate_contributors_list.sh` you should appear with proper name & email.

File: CONTRIBUTORS.txt [hunk 10]
```diff
@@ -152,11 +164,20 @@ needs to be listed here.
...
+- toluRV <111301355+toluRV@users.noreply.github.com>
```
Comment (weissi) [inline CONTRIBUTORS.txt:180]: CC @toluRV  if you want a real email address here, please submit a pull request with an appropriate entry to the `.mailmap` file. You'll find plenty of examples there, but essentially if you add

```
Your Name <your-real@email.address> <123456+the-weird-github-one@users.noreply.github.com>
```

and then run `scripts/generate_contributors_list.sh` you should appear with proper name & email.

**Pull request merged**